helix-core/src/main/java/org/apache/helix/messaging/handling/HelixTaskExecutor.java

```diff
@@ -734,6 +744,7 @@ void init() {
     }
 
     _isShuttingDown = false;
+    _isResetComplete = false;
```

Comment:
Why not put in onMessage? OnMessage is the place starting STs and changing the states.

Reply:
Correct me if I'm wrong. I think the problem is with `init()`, it initializes threadpools that need to be shutdown in reset(), so reset() should still be run even without state transition.
What `reset()` does should actually be shutdown() or close().